### PR TITLE
fix($parse): do not convert to string computed properties multiple times

### PR DESCRIPTION
**Pull request by lgalfaso**
Do not convert to string properties multiple times.

### CODE REVIEW

File: src/ng/parse.js [hunk 1]
```diff
@@ -1187,6 +1200,10 @@ ASTCompiler.prototype = {
     return 'ensureSafeFunction(' + item + ',text)';
   },
 
+  getStringValue: function(item) {
+    this.assign(item, 'getStringValue(' + item + ',text)');
```
Comment (petebacondarwin) [inline src/ng/parse.js]: should this be returning the result of the assign?

File: src/ng/parse.js [hunk 2]
```diff
@@ -38,20 +38,30 @@ var $parseMinErr = minErr('$parse');
...
 function ensureSafeMemberName(name, fullExpression) {
+  if (name === "__defineGetter__" || name === "__defineSetter__"
+      || name === "__lookupGetter__" || name === "__lookupSetter__"
+      || name === "__proto__") {
+    throw $parseMinErr('isecfld',
+        'Attempting to access a disallowed field in Angular expressions! '
+        + 'Expression: {0}', fullExpression);
+  }
+  return name;
+}
```
Comment (petebacondarwin) [inline src/ng/parse.js]: Have you checked that `ensureSafeMember` is never called with an unsafe name "thing"? I.E. That `getStringValue` is always called when needed?

File: src/ng/parse.js (hunk 1 above)
Comment (lgalfaso) [inline src/ng/parse.js]: the value is replaced inline, this is, `item` will have the new value.

File: test/ng/parseSpec.js [hunk 3]
```diff
@@ -2692,6 +2692,15 @@ describe('parser', function() {
           });
         });
 
+       it('should prevent the exploit', function() {
```
Comment (petebacondarwin) [inline test/ng/parseSpec.js]: Is this message purposefully vague right now?

File: src/ng/parse.js (hunk 2 above)
Comment (lgalfaso) [inline src/ng/parse.js]: ensureSafeMemberName is called in three cases.
* retrieving an identifier. Eg `foo`
* retrieving a non-computed member. Eg `foo.bar`
* retrieving a computed member. Eg `foo[bar]`

getStringValue is only used in the later. This is the expected behavior

File: test/ng/parseSpec.js (hunk 3 above)
Comment (lgalfaso) [inline test/ng/parseSpec.js]: If you have a better name, then I am ok with it. I do not know if the exploit can be explained in a one line.

**Comment by petebacondarwin** [comment]
Marvellous. LGTM - please merge.

**Comment by lgalfaso** [comment]
closed with 20cf7d5e3a0af766b1929e24794859c79439351c

**Comment by lgalfaso** [comment]
@petebacondarwin merged this at master and v1.4.x branch. Do you think this should this be backported to prior versions?

**Comment by petebacondarwin** [comment]
Yes.

**Comment by lgalfaso** [comment]
backport to 1.2 with afb65c11e5e3425f8431ad5b82308ff6b8ecbc64
backport to 1.3 with 9d5ac2eb84f21159ff93dac6f3d5782988a5f515

**Comment by petebacondarwin** [comment]
Good work!

**Pull request closed without merging**